### ISSUE TITLE
wazuh integration with gemini api

### ISSUE THREAD

**Issue by HEEEELLP** [wazuh/wazuh#25954]
### hello 
Can you help me with this? I’m trying to integrate the Gemini API with Wazuh to get an AI-generated description for alerts with a rule level of 15, but I’m not sure where the problem lies.

**/var/ossec/etc/ossec.conf**

```
<integration>
  <name>gemini.py</name>
  <rule_id>92213</rule_id>
  <level>15</level>
  <group>sysmon, sysmon_eid11_detections, windows</group>
  <event_location>/var/log/syslog</event_location>
  <alert_format>json</alert_format>
</integration>


```
### rules : 
```

<group name="gemini,">

    <rule id="92213" level="0">
        <decoded_as>json</decoded_as>
        <field name="integration">gemini</field>
        <description>gemini integration messages.</description>
        <options>no_full_log</options>
    </rule>

    <rule id="92213" level="3">
        <if_sid>92213</if_sid>
        <field name="gemini.error">904</field>
        <description>gemini: Error: Public API request rate limit reached</description>
        <options>no_full_log</options>
    </rule>

    <rule id="92213" level="7">
        <if_sid>92213</if_sid>
        <field name="gimini.result">906</field>
        <description>gemini: result - $(content_text)</description>
        <options>no_full_log</options>
    </rule>

</group>

```
### script python :
/var/ossec/integrations/gemini.py

```#!/var/ossec/framework/python/bin/python3
# Copyright (C) 2015-2023, Wazuh Inc.
# Gemini Integration

import json
import sys
import time
import os
from socket import socket, AF_UNIX, SOCK_DGRAM

try:
    import requests
except Exception as e:
    print("No module 'requests' found. Install: pip install requests")
    sys.exit(1)

# Global vars
debug_enabled = False
pwd = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))

json_alert = {}
now = time.strftime("%a %b %d %H:%M:%S %Z %Y")
# Set paths
log_file = '{0}/var/log/syslog'.format(pwd)
socket_addr = '{0}/queue/sockets/queue'.format(pwd)

def main(args):
    debug("# Starting")
    # Read args
    alert_file_location = args[1]
    apikey = args[2]
    hook_url = args[3]

    debug("# API Key")
    debug(apikey)
    debug("# File location")
    debug(alert_file_location)

    # Load alert. Parse JSON object.
    with open(alert_file_location) as alert_file:
        json_alert = json.load(alert_file)
    debug("# Processing alert")
    debug(json_alert)

    # Check for rule level 15
    if json_alert["rule"]["level"] == 15:
        debug("# Rule level 15 detected")
        # Send log to Gemini
        msg = request_gemini_info(json_alert, apikey, hook_url)
        # If positive match, send event to Wazuh Manager
        if msg:
            send_event(msg, json_alert["agent"])

def debug(msg):
    if debug_enabled:
        msg = "{0}: {1}\n".format(now, msg)
    print(msg)
    with open(log_file, "a") as f:
        f.write(str(msg))

def request_gemini_info(alert, apikey, hook_url):
    alert_output = {}
    
    # Gemini API request
    headers = {
        'Authorization': 'Bearer ' + apikey,
        'Content-Type': 'application/json',
    }

    json_data = {
        'full_log': alert['full_log'],
        'rule': {
            'id': alert['rule']['id'],
            'description': alert['rule']['description'],
            'level': alert['rule']['level']
        },
        'agent': alert['agent']
    }

    response = requests.post(hook_url, headers=headers, json=json_data)

    if response.status_code == 200:
        alert_output = response.json()
        debug("# Gemini API response received")
    else:
        alert_output["gemini"] = {}
        alert_output["integration"] = "custom-gemini"
        alert_output["gemini"]["error"] = response.status_code
        debug("# Error: The Gemini API encountered an error")
    
    return alert_output

def send_event(msg, agent=None):
    if not agent or agent["id"] == "000":
        string = '1:gemini:{0}'.format(json.dumps(msg))
    else:
        string = '1:[{0}] ({1}) {2}->gemini:{3}'.format(agent["id"], agent["name"], agent.get("ip", "any"), json.dumps(msg))

    debug(string)
    sock = socket(AF_UNIX, SOCK_DGRAM)
    sock.connect(socket_addr)
    sock.send(string.encode())
    sock.close()

if __name__ == "__main__":
    try:
        # Read arguments
        bad_arguments = False
        if len(sys.argv) >= 4:
            msg = '{0} {1} {2} {3}'.format(now, sys.argv[1], sys.argv[2], sys.argv[3])
            debug_enabled = (len(sys.argv) > 4 and sys.argv[4] == 'debug')
        else:
            msg = '{0} Wrong arguments'.format(now)
            bad_arguments = True

        # Logging the call
        with open(log_file, 'a') as f:
            f.write(str(msg) + '\n')

        if bad_arguments:
            debug("# Exiting: Bad arguments.")
            sys.exit(1)

        # Main function
        main(sys.argv)

    except Exception as e:
        debug(str(e))
        raise````

**Comment by mranv** [comment]
Latency is the key.

**Comment by mranv** [comment]
# Below are the list of issues you're facing. 

## 1) Integration config (ossec.conf)

* You’ve pinned `<rule_id>92213</rule_id>` while your stated goal is “AI description for all alerts with rule **level 15**.”
  ➜ With `<rule_id>` set, the integration will only run for alerts generated by that **single** rule (92213), not for all level=15 alerts.
  **Fix:** Remove `<rule_id>` or set it to the actual rule IDs you want, or rely solely on `<level>15</level>` (and optionally `<group>`).

* `<group>` filters to `sysmon, sysmon_eid11_detections, windows`, but rule 92213 (your “gemini integration messages”) won’t belong to those groups.
  **Fix:** If you want level=15 alerts from Sysmon to trigger the integration, keep the group filter and **remove `<rule_id>`**. If you want the integration’s own feedback events to be reprocessed by your custom rules, that’s a separate flow and shouldn’t be constrained by these sysmon groups.

* The integration will never fire if no incoming alerts **both** match the specified group(s) **and** have level 15 **and** have rule\_id 92213 (which they won’t).
  **Fix summary:** Use `<level>15</level>` (+ groups if you want), but drop `<rule_id>`.

## 2) Custom rules (duplicate IDs, typos, non-matching fields)

* **Duplicate rule IDs:** You used `id="92213"` for **three** rules. Wazuh rule IDs must be unique.
  **Fix:** Make them distinct, e.g.:

  * 92213: base “integration messages” rule
  * 92214: error rule (if\_sid 92213)
  * 92215: result rule (if\_sid 92213)

* **Typo in field name:**

  * You reference `<field name="gimini.result">906</field>` (typo: **gimini**).
  * Your Python sets `alert_output["gemini"]["error"]` on error, but on success you never create a `gemini.result` (or `content_text`) field.
    **Fix:** Standardize on `gemini.result` and actually put the AI text there (or reference the correct path in your rule).

* **Description placeholder mismatch:**

  * Rule uses `$(content_text)` but your message never provides `content_text`.
    **Fix:** Either change the rule to `$(gemini.result)` or have the script populate `content_text`.

* **Error code mismatch:**

  * Your rule checks `gemini.error = 904` but the script sets `gemini.error = response.status_code` (e.g., 401/403/429).
    **Fix:** Either normalize error codes to your custom values (e.g., map 429→904) or match real HTTP codes in the rules.

* **Group name punctuation:**

  * `<group name="gemini,">` includes a trailing comma; that’s unusual and can cause confusion.
    **Fix:** Use `<group name="gemini">`.

## 3) Python script (API usage, output shape, logging paths)

* **Gemini API auth is wrong (unless you have a proxy):**

  * You’re using `Authorization: Bearer <key>` and `json` payload in a shape Gemini doesn’t accept.
  * The public Gemini REST API expects `?key=API_KEY` in the URL and a `generateContent` body shape (`contents: [{parts: [{text: "..."}]}]`), or OAuth Bearer when using **Vertex AI**, not a raw “full\_log/rule/agent” JSON.
    **Fix:**
  * If you’re calling Google’s **public Generative Language API** directly, format the URL and JSON as per `generateContent` and pass `?key=` (not Bearer).
  * If you’re calling your own **hook\_url** (middleware), then **your middleware** must translate your JSON into a proper Gemini request and return a uniform response that your rules expect (e.g., `{"integration":"gemini","gemini":{"result":"..."} }`).

* **You never set a success field your rules match on:**

  * On success, you do `alert_output = response.json()` and send it back, but your rules look for `gemini.result` or `content_text`.
    **Fix:** After getting the response, **extract the model’s text** and set a predictable field, e.g.:

  ```python
  alert_output = {
      "integration": "gemini",
      "gemini": {
          "result": "<the generated text>"
      }
  }
  ```

  Then align your rule’s `<field name="gemini.result">` or use the variable `$(gemini.result)` in descriptions.

* **Log file path likely wrong/permission issues:**

  * `log_file = '{pwd}/var/log/syslog'` will resolve to `/var/ossec/var/log/syslog` (or similar), which is **not** the system’s `/var/log/syslog` and may not exist/be writable by the Wazuh user.
    **Fix:** Use Wazuh’s integrations log (e.g., `/var/ossec/logs/integrations.log`) or a known writable path inside Wazuh.

* **Socket path seems fine, but verify:**

  * `socket_addr = '{pwd}/queue/sockets/queue'` is standard for Wazuh manager; just ensure the script runs on the **manager** and has permissions.

* **CLI arguments contract:**

  * You assume: `alert_file`, `apikey`, `hook_url`. Ensure ossec.conf provides `<api_key>` and `<hook_url>` entries that Wazuh passes to your script in that order (this is how official integrations do it).
    **Fix:** Add to `<integration>`:

  ```xml
  <api_key>YOUR_API_KEY</api_key>
  <hook_url>https://.../v1beta/models/...:generateContent?key=...</hook_url>
  ```

  and confirm the order Wazuh passes them (alert file is first). If using Vertex/OAuth, don’t use `<api_key>`—use your middleware or service account/OAuth flow.

* **Missing shebang & exec perms:**

  * No `#!/usr/bin/env python3` and file may not be executable.
    **Fix:** Add shebang and `chmod +x /var/ossec/integrations/gemini.py`.

## 4) Event flow confusion (who triggers what)

* Right now, you’re trying to both **trigger** on level 15 alerts and also **create** new “gemini integration messages” with the same rule id. That loops logic.
  **Recommended flow:**

  1. Integration triggers on **original** alerts (e.g., any alert with `level >= 15`, and maybe certain groups).
  2. Script calls Gemini, builds a normalized JSON with `integration: "gemini"` and `gemini.result: "<text>"`.
  3. You have **separate rules** (distinct IDs) that match those returned integration events to display/forward the AI summary.

---

# Minimal set of changes to unblock

1. **ossec.conf**

```xml
<integration>
  <name>gemini.py</name>
  <level>15</level>
  <group>sysmon, sysmon_eid11_detections, windows</group>
  <alert_format>json</alert_format>
  <api_key>YOUR_API_KEY (or remove if using Vertex/OAuth)</api_key>
  <hook_url>YOUR_ENDPOINT (real Gemini URL with ?key=... or your middleware)</hook_url>
</integration>
```

(Notice: **removed** `<rule_id>`.)

2. **rules (unique IDs, fixed fields)**

```xml
<group name="gemini">
  <rule id="92213" level="0">
    <decoded_as>json</decoded_as>
    <field name="integration">gemini</field>
    <description>gemini integration messages.</description>
    <options>no_full_log</options>
  </rule>

  <rule id="92214" level="3">
    <if_sid>92213</if_sid>
    <field name="gemini.error">429</field>  <!-- or whatever you actually set -->
    <description>gemini: Error: Public API request rate limit reached</description>
    <options>no_full_log</options>
  </rule>

  <rule id="92215" level="7">
    <if_sid>92213</if_sid>
    <field name="gemini.result">.+</field> <!-- presence check -->
    <description>gemini: result - $(gemini.result)</description>
    <options>no_full_log</options>
  </rule>
</group>
```

3. **gemini.py (normalize output + fix logging)**

* Use a writable log, e.g. `/var/ossec/logs/integrations.log`.
* After a successful call, extract text and populate `gemini.result`.
* Map HTTP 429→`gemini.error=429` (or your custom code, but be consistent with rules).

Example success shaping (inside `request_gemini_info` after a 200):

```python
data = response.json()
# Extract the model text depending on your hook:
generated = extract_text(data)  # implement this per your endpoint
alert_output = {
  "integration": "gemini",
  "gemini": {"result": generated},
}
```

---

## TL;DR

* Remove `<rule_id>` from the integration so it actually triggers on **all** level 15 alerts in those groups.
* Fix duplicate rule IDs, typos (`gimini` → `gemini`), and make rules match the **real** fields you emit.
* Either call Gemini correctly (public API with `?key=` and proper body) or have your hook translate — but in any case, return a **normalized** payload with `integration:"gemini"` and `gemini.result:"..."`.
* Fix logging path and add a shebang/exec perms.
